File: python/xor_cipher.py
import random


class Pair:
    """
    A class to represent a pair of key and cypher.

    Attributes
    ----------
    key : list of int
        The key used for encryption and decryption.
    cypher : bytearray
        The encrypted message.
    """

    def __init__(self, key, cypher):
        """
        Constructs all the necessary attributes for the Pair object.

        Parameters
        ----------
        key : list of int
            The key used for encryption and decryption.
        cypher : bytearray
            The encrypted message.
        """
        self.key = key
        self.cypher = cypher


def encrypt(message):
    """
    Encrypts a message using a random key.

    Parameters
    ----------
    message : str
        The message to be encrypted.

    Returns
    -------
    Pair
        A Pair object containing the key and the cypher.
    """
    content = message.encode('utf-8')
    cypher = bytearray(len(content))
    key = [random.randint(0, 1) for _ in range(len(content))]

    for i in range(len(content)):
        cypher[i] = content[i] ^ key[i]

    return Pair(key, cypher)


def decrypt(pair):
    """
    Decrypts a cypher using the given key.

    Parameters
    ----------
    pair : Pair
        A Pair object containing the key and the cypher.

    Returns
    -------
    str
        The decrypted message.
    """
    message = bytearray(len(pair.cypher))

    for i in range(len(pair.cypher)):
        message[i] = pair.cypher[i] ^ pair.key[i]

    return message.decode('utf-8')


def main():
    """
    Main function to demonstrate encryption and decryption.
    """
    message = "Here is the text."

    print("Encrypting:", message)
    p = encrypt(message)

    print("Key:", p.key)
    print("Cypher:", list(p.cypher))
    print("-------------------------")
    print("Decrypted Message:", decrypt(p))


if __name__ == "__main__":
    main()

File: java/XorCipher.java
import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.concurrent.ThreadLocalRandom;

/**
 * A class to perform XOR encryption and decryption.
 */
public class XorCipher {
    /**
     * The main method to demonstrate encryption and decryption.
     *
     * @param args command-line arguments (not used)
     */
    public static void main(String[] args) {
        final String message = "Here is the text.";

        final Pair p = encrypt(message);

        System.out.println(Arrays.toString(p.key()));
        System.out.println(Arrays.toString(p.cypher()));

        System.out.println(decrypt(p));
    }

    /**
     * Encrypts a message using a randomly generated key.
     *
     * @param message the message to be encrypted
     * @return a Pair containing the encryption key and the encrypted message
     */
    public static Pair encrypt(final String message) {
        final byte[] content = message.getBytes(StandardCharsets.UTF_8);
        final byte[] cypher = new byte[content.length];
        final int[] key = key(content.length);

        for (int i = 0; i < content.length; i++) {
            final byte b = content[i];
            cypher[i] = (byte) (b ^ key[i]);
        }

        return new Pair(key, cypher);
    }

    /**
     * Decrypts a message using the provided key.
     *
     * @param pair a Pair containing the encryption key and the encrypted message
     * @return the decrypted message as a String
     */
    public static String decrypt(final Pair pair) {
        final byte[] message = new byte[pair.cypher().length];

        for (int i = 0; i < pair.cypher().length; i++) {
            final int key = pair.key()[i];
            final int c = pair.cypher()[i];
            message[i] = (byte) (key ^ c);
        }

        return new String(message, StandardCharsets.UTF_8);
    }

    /**
     * Generates a random key of the specified length.
     *
     * @param length the length of the key to be generated
     * @return an array of integers representing the key
     */
    public static int[] key(int length) {
        final int[] key = new int[length];

        for (int i = 0; i < length; i++) {
            key[i] = ThreadLocalRandom.current().nextInt(2);
        }

        return key;
    }

    /**
     * A record to hold the encryption key and the encrypted message.
     *
     * @param key    an array of integers representing the key
     * @param cypher an array of bytes representing the encrypted message
     */
    public static record Pair(int[] key, byte[] cypher) {
    }
}